org.eclipse.xtext.ui.testing/src/org/eclipse/xtext/ui/testing/AbstractCodeMiningTest.java

```diff
@@ -0,0 +1,123 @@
+/*******************************************************************************
+ * Copyright (c) 2020 itemis AG (http://www.itemis.eu) and others.
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ * 
+ * SPDX-License-Identifier: EPL-2.0
+ *******************************************************************************/
+package org.eclipse.xtext.ui.testing;
+
+import java.util.List;
+import java.util.concurrent.ExecutionException;
+
+import org.eclipse.core.resources.IFile;
+import org.eclipse.core.runtime.NullProgressMonitor;
+import org.eclipse.jface.text.codemining.ICodeMining;
+import org.eclipse.jface.text.codemining.ICodeMiningProvider;
+import org.eclipse.jface.text.codemining.LineHeaderCodeMining;
+import org.eclipse.jface.text.source.ISourceViewer;
+import org.eclipse.xtext.resource.FileExtensionProvider;
+import org.eclipse.xtext.ui.editor.XtextEditor;
+
+import com.google.common.annotations.Beta;
+import com.google.inject.Inject;
+
+/**
+ * @author miklossy - Initial contribution and API
+ * @since 2.23
+ */
+@Beta
+public class AbstractCodeMiningTest extends AbstractEditorTest {
+
+	@Inject protected FileExtensionProvider fileExtensionProvider;
+	@Inject protected ICodeMiningProvider codeMiningProvider;
+
+	/**
+	 * Test that the expected code minings are present on a given DSL text.
+	 *
+	 * @param The initial DSL text.
+	 * @param expected The DSL text where the expected code minings are inserted on the right position.
+	 */
+	public void testCodeMining(CharSequence initialText, String expected) {
+		// Given
+		IFile dslFile = dslFile(initialText);
+		// When
+		XtextEditor editor = openEditor(dslFile);
+		// Then
+		codeMiningsArePresent(editor, expected);
+	}
+
+	@Override
+	protected XtextEditor openEditor(IFile dslFile) {
+		try {
+			XtextEditor editor = super.openEditor(dslFile);
+			assertNotNull(editor);
+			return editor;
+		} catch (Exception e) {
+			throw new RuntimeException(e);
+		}
+	}
+
+	protected void codeMiningsArePresent(XtextEditor editor, String expected) {
+		String actual = insertCodeMinings(editor);
+		assertEquals(expected, actual);
+	}
+
+	protected String insertCodeMinings(XtextEditor editor) {
+		ISourceViewer viewer = editor.getInternalSourceViewer();
+
+		String text = editor.getDocument().get();
+		StringBuilder sb = new StringBuilder();
+
+		List<? extends ICodeMining> codeMinings;
+		try {
+			codeMinings = codeMiningProvider.provideCodeMinings(viewer, new NullProgressMonitor()).get();
+		} catch (InterruptedException | ExecutionException e) {
+			throw new RuntimeException(e);
+		}
+
+		int currentOffset = 0;
+
+		for (ICodeMining codeMining : codeMinings) {
```

Comment:
does this work with multiple codeminings? does it work with multiple codeminings at the same place? does it work if the codeminings are not sorted by offset?

Reply:
Multiple codeminings are already supported, see org.eclipse.xtext.example.arithmetics.ui.tests.codemining.CodeMiningTest test class in the PR:
![screenshot](https://user-images.githubusercontent.com/20393472/89009105-3033f080-d30c-11ea-856a-4c41b37fadca.png)

I think, the other corner cases you mentioned are not supported, but the user has the possibility to override the `protected String insertCodeMinings(XtextEditor editor)` method to provide a custom implementation for such cases.

Reply:
i dont think we should let that up to the user ....

Reply:
the usual algorithm is to sort codeminings by offset backwards (apply them from bottom to top)

Reply:
e.g. adding 

```
		for (Property f :  EcoreUtil2.eAllOfType(resource.getContents().get(0), Property.class)) {
			int endOffset = NodeModelUtils.findActualNodeFor(f).getEndOffset();
			acceptor.accept(createNewLineContentCodeMining(endOffset, " Breaking Tamas Test"));
		}
```

to domain model code mining provider easily breaks your code

Reply:
done